realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyMediatorGenerator.java

```diff
@@ -0,0 +1,333 @@
+        writer.emitEmptyLine();
+    }
+
+    private void emitCopyToRealmMethod(JavaWriter writer) throws IOException {
+        writer.emitAnnotation("Override");
+        writer.beginMethod(
+                "<E extends RealmObject> E",
+                "copyOrUpdate",
+                EnumSet.of(Modifier.PUBLIC),
+                "Realm", "realm", "E", "obj", "boolean", "update", "Map<RealmObject, RealmObjectProxy>",  "cache"
+        );
+
+        writer.emitStatement("Class<E> clazz = (Class<E>) ((obj instanceof RealmObjectProxy) ? obj.getClass().getSuperclass() : obj.getClass())");
+        writer.emitEmptyLine();
+        emitMediatorSwitch(new ProxySwitchStatement() {
+            @Override
+            public void emitStatement(int i, JavaWriter writer) throws IOException {
+                writer.emitStatement("return (E) %s.copyOrUpdate(realm, (%s) obj, update, cache)", proxyClasses.get(i), simpleModelClasses.get(i));
+            }
+        }, writer, false);
+        writer.endMethod();
+        writer.emitEmptyLine();
+    }
+
+    private void emitCreteOrUpdateUsingJsonObject(JavaWriter writer) throws IOException {
+        writer.emitAnnotation("Override");
+        writer.beginMethod(
+                "<E extends RealmObject> E",
+                "createOrUpdateUsingJsonObject",
+                EnumSet.of(Modifier.PUBLIC),
+                Arrays.asList("Class<E>", "clazz", "Realm", "realm", "JSONObject", "json", "boolean", "update"),
+                Arrays.asList("JSONException")
+        );
+        emitMediatorSwitch(new ProxySwitchStatement() {
+            @Override
+            public void emitStatement(int i, JavaWriter writer) throws IOException {
+                writer.emitStatement("return (E) %s.createOrUpdateUsingJsonObject(realm, json, update)", proxyClasses.get(i));
+            }
+        }, writer);
+        writer.endMethod();
+        writer.emitEmptyLine();
+    }
+
+    private void emitCreateUsingJsonStream(JavaWriter writer) throws IOException {
+        writer.emitAnnotation("Override");
+        writer.beginMethod(
+                "<E extends RealmObject> E",
+                "createUsingJsonStream",
+                EnumSet.of(Modifier.PUBLIC),
+                Arrays.asList("Class<E>", "clazz", "Realm", "realm", "JsonReader", "reader"),
+                Arrays.asList("java.io.IOException")
+        );
+        emitMediatorSwitch(new ProxySwitchStatement() {
+            @Override
+            public void emitStatement(int i, JavaWriter writer) throws IOException {
+                writer.emitStatement("return (E) %s.createUsingJsonStream(realm, reader)", proxyClasses.get(i));
+            }
+        }, writer);
+        writer.endMethod();
+        writer.emitEmptyLine();
+    }
+
+    // Emits the control flow for selecting the appropriate proxy class based on the model class
+    // Currently it is just if..else, which is inefficient for large amounts amounts of model classes.
```

Comment:
Please share the experimental code. I'm sure this can be done more elegantly & faster. The generated code also takes up some space, which we increasingly have to consider.